File: lib/clients/dbus/base/abstract_dbus_client.py
from abc import ABC
from typing import Any, Callable

# pylint: disable=E0611
from PyQt5.QtCore import QObject, pyqtSlot
from PyQt5.QtDBus import QDBusInterface, QDBusConnection, QDBusReply, QDBusMessage, QDBusServiceWatcher

from lib.utils.event_bus import event_bus
from lib.utils.logger import Logger
from lib.utils.singleton import singleton


@singleton
class PropertyChangeListener(QObject):
    """Class for properties change listening"""

    def __init__(self):
        super().__init__()
        self.watched_ifaces = []
        self.sys_bus = QDBusConnection.systemBus()
        self.ses_bus = QDBusConnection.systemBus()
        self.sys_bus.connect(
            "",
            "",
            "org.freedesktop.DBus.Properties",
            "PropertiesChanged",
            self.on_properties_changed,
        )
        self.ses_bus.connect(
            "",
            "",
            "org.freedesktop.DBus.Properties",
            "PropertiesChanged",
            self.on_properties_changed,
        )

    @pyqtSlot(QDBusMessage)
    def on_properties_changed(self, message: QDBusMessage):
        """Slot for signal"""
        args = message.arguments()
        if len(args) < 3:
            return

        interface_name, changed_properties, _ = args
        if interface_name in self.watched_ifaces:
            for prop, value in changed_properties.items():
                event_bus.emit(f"{interface_name}.{prop}", value)

    def add_interface(self, iface: str):
        """Add interface to watch for"""
        if not iface in self.watched_ifaces:
            self.watched_ifaces.append(iface)


propertyChangeListener = PropertyChangeListener()


class AbstractDbusClient(ABC):
    """Base class for dbus clients"""

    def __init__(self, system_bus: bool, service_name: str, object_path: str, interface_name: str):
        self._logger = Logger(self.__class__.__name__)
        bus = QDBusConnection.systemBus() if system_bus else QDBusConnection.sessionBus()

        self._interface = QDBusInterface(service_name, object_path, interface_name, bus)
        if not self._interface.isValid():
            raise Exception(
                f"Not found {"system" if system_bus else "session"} DBus service {service_name} with path {object_path} and interface {interface_name}"  # pylint: disable=C0301
            )
        self._props_interface = QDBusInterface(service_name, object_path, "org.freedesktop.DBus.Properties", bus)

        self._available = True
        self._service_name = service_name
        self._interface_name = interface_name

        self.watcher = QDBusServiceWatcher(
            service_name,
            bus,
            QDBusServiceWatcher.WatchForOwnerChange,
        )
        self.watcher.serviceOwnerChanged.connect(self._on_service_owner_changed)

        propertyChangeListener.add_interface(interface_name)

    def _on_service_owner_changed(self, service_name, old_owner, new_owner):
        if service_name == self._service_name:
            if not new_owner:
                self._available = False
                self._logger.warning(f"Service '{service_name}' is not available")
            elif not old_owner:
                self._available = True
                self._logger.info(f"Service '{service_name}' is available ")

    def on(self, prop_name, function: Callable[[Any], None]):
        """Subscribe to signal"""
        event_bus.on(f"{self._interface_name}.{prop_name}", function)

    def _check_availability(self):
        if not self._available:
            raise Exception(f"Service {self._service_name} is not available")

    def _get_property(self, property_name: str):
        """Get dbus property"""
        self._check_availability()
        reply = QDBusReply(self._props_interface.call("Get", self._interface_name, property_name))
        if reply.isValid():
            return reply.value()

        raise Exception(f"Error getting property {property_name}: {reply.error().message()}")

    def _set_property(self, property_name: str, value):
        """Set dbus property"""
        self._check_availability()
        reply = QDBusReply(self._props_interface.call("Set", self._interface_name, property_name, value))
        if not reply.isValid():
            raise Exception(f"Error setting property {property_name}: {reply.error().message()}")

    def _invoke_method(self, method_name: str, *args):
        """Invoke dbus method"""
        self._check_availability()
        reply = QDBusReply(self._interface.call(method_name, *args))
        if reply.isValid():
            return reply.value()

        raise Exception(f"Error invoking method {method_name}: {reply.error().message()}")

File: RogControlCenter/src/clients/dbus/abstract/abstract_dbus_client.cpp

#include "../../../../include/clients/dbus/abstract/abstract_dbus_client.hpp"

#include <QtDBus/QDBusMessage>
#include <QtDBus/QDBusReply>
#include <QtDBus/QDBusVariant>

AbstractDbusClient::AbstractDbusClient(bool systemBus, const QString& service, const QString& objectPath, const QString& interface, bool required,
									   QObject* parent)
	: QObject(parent),
	  systemBus_(systemBus),
	  serviceName_(service),
	  objectPath_(objectPath),
	  interfaceName_(interface),
	  bus_(systemBus ? QDBusConnection::systemBus() : QDBusConnection::sessionBus()),
	  iface_(nullptr),
	  available_(false) {
	if (!bus_.isConnected()) {
		if (required) {
			throw std::runtime_error("Failed to connect to D-Bus");
		}
		return;
	}

	iface_	   = new QDBusInterface(serviceName_, objectPath_, interfaceName_, bus_, this);
	available_ = iface_->isValid();
	if (!available_) {
		if (required) {
			throw std::runtime_error("Failed to create D-Bus interface: " + serviceName_.toStdString());
		}
	}

	// === Conectar la señal de cambios de propiedades ===
	// Escucha cualquier cambio en las propiedades de la interfaz
	bus_.connect(serviceName_,												   // service
				 objectPath_,												   // path
				 "org.freedesktop.DBus.Properties",							   // interfaz que emite la señal
				 "PropertiesChanged",										   // señal a escuchar
				 this,														   // objeto que recibirá la señal
				 SLOT(onPropertiesChanged(QString, QVariantMap, QStringList))  // slot a ejecutar
	);
}

AbstractDbusClient::~AbstractDbusClient() = default;

bool AbstractDbusClient::available() const {
	return available_;
}

void AbstractDbusClient::checkAvailable() const {
	if (!available_) {
		throw std::runtime_error("D-Bus service not available: " + serviceName_.toStdString());
	}
}